daml/ContingentClaims/Claim.daml

```diff
@@ -23,45 +22,45 @@ type F = ClaimF
 -- See [quickstart](../QUICKSTART.md) for detailed explanation.
 -- In Petyon-Jones' paper, this is called 'Contract'.
 -- We renamed it to avoid ambiguity.
--- * `f`, `t` and `x` respectively correspond to the `Observable`, it's input type and the resulting output type.
+-- * `t` and `x` respectively correspond to the `Observation`s, input type and the resulting output type.
 -- * `a` is the representation of an asset, e.g. a `Text` ISIN code.
-data Claim f t x a
+data Claim t x a
   = Zero
       -- ^ Represents an absence of claims.
   | One a
       -- ^ The bearer acquires one unit of `a` *immediately*.
-  | Give (Claim f t x a)
+  | Give (Claim t x a)
       -- ^ The obligations of the bearer and issuer are revesed.
-  | And with lhs: Claim f t x a, rhs: Claim f t x a
+  | And with fst: Claim t x a, snd: Claim t x a, tail: [Claim t x a]
```

Comment:
Thinking about the usability of this representation, I'm not entirely convinced this is the best way to enforce this. it prevents us from folding or mapping over `And`s and makes it clunkier to work with it in general. Could we not add a validation function that checks the minimum elements? I also think we could allow `And` with a single element, and it would simplify to just this element instead. And an `And` with zero elements could essentially be pruned off the tree. Just some thoughts, not sure what the best options is here, but worth discussing.

Reply:
I agree with you that this doesn't feel very nice.

>  it prevents us from folding or mapping over Ands 

That's not quite right, unless I've misunderstood what you're saying:
```
instance Functor Claim where
  fmap f (And fst snd tail) = And (f fst) (f snd) (fmap f tail)
```
and ditto for `fold`

> Could we not add a validation function that checks the minimum elements?

The problem is we can't use `ensure` because a `Claim` is not a `template`. So there is no way to enforce it at the data type level.

I mentioned using "smart constructors". I'll look into that again.
Another idea I had was adding a `monoid` instance, so that you can just `And` together multiple claims using the `<>` operator.

Reply:
Re: smart constructors, running into a problem we had last time, raised a bug for it now: https://github.com/digital-asset/daml/issues/12166

Reply:
Also, I'm consider whether this alternative representation would be appropriate, since we're already messing around with this:

```
And inits : [ Claim] , penultimate : Claim, last : Claim
```

I _think_ this might be considerably more efficient than the current approach. Need to mull over it a bit more.